WalletWasabi/Blockchain/Analysis/BlockchainAnalyzer.cs

```diff
@@ -3,60 +3,182 @@
+				anonset -= indistinguishableWalletOutputs[newCoin.Amount] - 1;
 
+				HdPubKey hdPubKey = newCoin.HdPubKey;
+				if (hdPubKey.AnonymitySet == HdPubKey.DefaultHighAnonymitySet)
+				{
+					// If the new coin's HD pubkey haven't been used yet
+					// then its anonset haven't been set yet.
+					// In that case the acquired anonset does not have to be intersected with the default anonset,
+					// so this coin gets the aquired anonset.
+					hdPubKey.AnonymitySet = anonset;
+				}
+				else if (distinctWalletInputPubKeys.Contains(hdPubKey))
+				{
+					// If it's a reuse of an input's pubkey, then intersection punishment is senseless.
+					hdPubKey.AnonymitySet = inheritedAnonset + 1;
+				}
+				else if (tx.WalletOutputs.Where(x => x != newCoin).Select(x => x.HdPubKey).Contains(hdPubKey))
+				{
+					// If it's a reuse of another output' pubkey, then intersection punishment can only go as low as the inherited anonset.
+					hdPubKey.AnonymitySet = Math.Max(inheritedAnonset + 1, Intersect(new[] { anonset, hdPubKey.AnonymitySet }, 1));
+				}
+				else
+				{
+					hdPubKey.AnonymitySet = Intersect(new[] { anonset, hdPubKey.AnonymitySet }, 1);
+				}
+			}
+		}
+
+		private void AnalyzeWalletInputs(SmartTransaction tx, out HashSet<HdPubKey> distinctWalletInputPubKeys, out int inheritedAnonset)
+		{
+			var ownInputCount = tx.WalletInputs.Count;
+			distinctWalletInputPubKeys = tx.WalletInputs.Select(x => x.HdPubKey).ToHashSet();
+			inheritedAnonset = 0;
+			if (ownInputCount > 0)
+			{
 				// If we provided inputs to the transaction.
-				if (tx.WalletInputs.Any())
+				// Reusing pubkey on the input side is good, the punishment happened already.
+				var distinctWalletInputPubKeyCount = distinctWalletInputPubKeys.Count;
+				var pubKeyReuseCount = ownInputCount - distinctWalletInputPubKeyCount;
+				var privacyBonus = Intersect(distinctWalletInputPubKeys.Select(x => x.AnonymitySet), (double)distinctWalletInputPubKeyCount / (tx.Transaction.Inputs.Count - pubKeyReuseCount));
+
+				// If the privacy bonus is <=1 then we are not inheriting any privacy from the inputs.
+				var normalizedBonus = privacyBonus - 1;
+
+				// And add that to the base anonset from the tx.
+				inheritedAnonset = normalizedBonus;
+
+				foreach (var key in tx.WalletInputs.Select(x => x.HdPubKey))
 				{
-					// Take the input that we provided with the smallest anonset.
-					// And add that to the base anonset from the tx.
-					// Our smallest anonset input is the relevant here, because this way the common input ownership heuristic is considered.
-					// Take minus 1, because we do not want to count own into the anonset, so...
-					// If the anonset of our UTXO would be 1, and the smallest anonset of our inputs would be 1, too, then we don't make...
-					// The new UTXO's anonset 2, but only 1.
-					anonset += tx.WalletInputs.Select(x => x.HdPubKey).Min(x => x.AnonymitySet) - 1;
+					key.AnonymitySet = inheritedAnonset + 1;
```

Comment:
I removed 90% of +-1 thingy by changing the concept: instead of giving back inherited anonset we give back new input anonset, so most of the +-1 isn't necessary anymore. [`5ae96fa` (#4724)](https://github.com/zkSNACKs/WalletWasabi/pull/4724/commits/5ae96fa3d73d5ebf4c68168c0982309f88716755)

Reply:
ACK